sound/soc/sof/intel/cnl.c

```diff
@@ -150,14 +151,77 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
 				CNL_DSP_REG_HIPCCTL_DONE);
 }
 
+static bool cnl_pm_gate_compress(struct snd_sof_ipc_msg *msg,
+				 u32 *dr, u32 *dd)
+{
+	struct sof_ipc_pm_gate *pm_gate;
+
+	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE)) {
+		pm_gate = msg->msg_data;
+
+		/* send the compact message via the primary register */
+		*dr = HDA_IPC_MSG_COMPACT | HDA_IPC_PM_GATE;
+
+		/* send payload via the extended data register */
+		*dd = pm_gate->flags;
+
+		return true;
+	}
+
+	return false;
+}
+
+static const struct hda_ipc compact_ipc[] = {
+	{.header = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE,
+	 .compress = cnl_pm_gate_compress},
+};
+
+static bool cnl_is_ipc_compact(struct snd_sof_ipc_msg *msg)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(compact_ipc); i++)
+		if (msg->header == compact_ipc[i].header)
+			return true;
+
+	return false;
+}
+
+static bool cnl_compact_ipc_compress(struct snd_sof_ipc_msg *msg,
+				     u32 *dr, u32 *dd)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(compact_ipc); i++)
+		if (msg->header == compact_ipc[i].header)
+			return compact_ipc[i].compress(msg, dr, dd);
```

Comment:
I really don't see the point of all this abstraction. At the end of the day, you are still dependent on register definitions (dd, dr) that only work for cAVS 1.8.
This looks like a solution that's either not complete or is half-baked.

Reply:
As the compact IPC itself could be very IPC specific, with respect to how to fill these dr and dd registers, so each IPC could have it's specific function to fill them, e.g.:

PM_GATE: header->dr, flags->dd;
PM_CONTEXT_SAVE: header->dr, nothing->dd;
ABI_QUERY: header->dr, driver ABI compress(Major Bytes << 24 || Minor Bytes << 16 || Revision) -> dd.
...

In the extreme case, we can compress all our IPCs to compact ones in platform where there is no mailbox(e.g. suecreek?).

Using switch in a combined compress() function is also fine, except that the functional itself may become longer and longer and the handling for different kind of IPC structure may be very annoying there, imagine this:
```
static bool cnl_ipc_compress(struct snd_sof_ipc_msg *msg,
				 u32 *dr, u32 *dd)
{
	struct sof_ipc_pm_gate *pm_gate; /* Keyon: what structure should be define here? */

	switch (msg->header) {
case SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE:
		pm_gate = msg->msg_data;

		/* send the compact message via the primary register */
		*dr = HDA_IPC_MSG_COMPACT | HDA_IPC_PM_GATE;

		/* send payload via the extended data register */
		*dd = pm_gate->flags;

		return true;
	       break;
        }
	return false;
}
```